cnxarchive/views/xpath.py

```diff
@@ -0,0 +1,147 @@
+    IdentHashShortId, IdentHashMissingVersion,
+    split_ident_hash, COLLECTION_MIMETYPE
+    )
+from .helpers import get_uuid, get_latest_version
+
+
+@view_config(route_name='xpath-book', request_method='GET')
+def xpath_book(request):
+    """
+    Given a book UUID (and optional version) and escaped xpath string:
+
+    if Accept: 'application/xhtml+xml' is in the headers, returns the HTML tree of the book, with each module title a
+    link which returns the xpath query result for that page UUID (and version if supplied, else uses most recent), or
+
+    otherwise, returns a JSON object of results, each result containing:
+    module_name,
+    module_uuid,
+    xpath_results, an array of strings, each an individual xpath result.
+    """
+
+    if 'application/xhtml+xml' in request.headers.get('ACCEPT', ''):
+        return xpath_book_html(request)
+    else:
+        return execute_xpath(request, 'xpath')
+
+
+@view_config(route_name='xpath-page', request_method='GET', renderer='json')
+def xpath_page(request):
+    """Given a page UUID (and optional version), returns a JSON object of results, as in xpath_book()"""
+    return execute_xpath(request, 'xpath-module')
+
+
+def xpath_book_html(request):
+    result = _get_content_json()
+
+    if result['mediaType'] == COLLECTION_MIMETYPE:
+        content = tree_to_html_xpath(result['tree'])
+    else:
+        exc = httpexceptions.HTTPBadRequest()
+        exc.explanation = 'Received non-book UUID'
+        raise exc
+
+    resp = request.response
+    resp.status = "200 OK"
+    resp.content_type = 'application/xhtml+xml'
+    resp.body = content
+    return resp
+
+
+def tree_to_html_xpath(tree):
+    """Return html list version of book tree."""
+    ul = etree.Element('ul')
+    html_listify_xpath([tree], ul)
+    return HTML_WRAPPER.format(etree.tostring(ul))
+
+
+def html_listify_xpath(tree, root_ul_element, parent_id=None):
+    """Recursively construct HTML nested list version of book tree, adding hrefs to the page titles which are links to
+    the the API endpoint which runs the xpath on that page CNXML.
+    The original caller should not call this function with the
+    `parent_id` defined.
+
+    """
+    request = get_current_request()
+    is_first_node = parent_id is None
+    if is_first_node:
+        parent_id = tree[0]['id']
+    for node in tree:
+        li_elm = etree.SubElement(root_ul_element, 'li')
+        a_elm = etree.SubElement(li_elm, 'a')
+        a_elm.text = node['title']
+        if node['id'] != 'subcol':
+            if is_first_node:
+                a_elm.set('href', request.route_path(
+                    # FIXME is there a prettier way to do this?
+                    'xpath-book', ident_hash=node['id']) + "?q=" + request.params.get('q', ''))
+            else:
+                a_elm.set('href', request.route_path(
+                    'xpath-page',
+                    ident_hash=node['id']) + "?q=" + request.params.get('q', ''))
```

Comment:
Do you mean like [this?](https://archive.cnx.org/contents/185cbf87-c72e-48f5-b51e-f14f21b5eabd%4010.117.html) Perhaps I misunderstood what you were asking for at the beginning. I was aiming at creating an HTML tree that lists links to getting the xpath results from a page. One use case is our intern project - from the frontend, I will retrieve this tree with `xpath=//*[local-name()=\'definition\']`, then allow the users to select links which will return definitions (skipping the backend step, which will be a proxy, passing the request from the frontend to cnx-archive, then parsing the result into Card objects).